### PR TITLE
Refactor test parameter validation into shared helpers

### LINKED ISSUE #1572: Test Generalization and Helper Functions
### Is your feature request related to a problem? Please describe.

Currently, many of our tests contain duplicated code for common operations. While we have created some reusable helpers in src/server/test/util/validationHelpers.js, there are additional opportunities to generalize test patterns. 

### Describe the solution you'd like

Refactor existing tests to use the helper functions from src/server/test/util/validationHelpers.js more consistently, and create additional helper functions for common patterns that aren't yet covered.

### Describe alternatives you've considered
Leave tests as-is - Continue with duplicated code. This is easier short-term but harder to maintain long-term.

### Additional context

Available helper functions in src/server/test/util/validationHelpers.js:
  - validateString() - validates string fields with min/max length and enum constraints
  - validateInt() - validates integer fields with bounds
  - validateBool() - validates boolean fields
  - testInvalidField() - tests individual field validation
  - validateToken() - tests token validation
  - validateNoExtraFields() - validates rejection of extra parameters
  - validateCommaSeparatedIdPatterns() - validates ID list formats
  - expectValidCommaSeparatedIds() - verifies valid ID lists

Reference: PR #1528 in this [comment](https://github.com/OpenEnergyDashboard/OED/pull/1528#pullrequestreview-3479162419)

### PR DESCRIPTION
**Pull request by KaulanSerzhanuly**
# Description

- Refactors multiple `ParamsTest.js` files to use shared helpers from src/server/test/util/validationHelpers.js instead of duplicating validation logic.
- Adds/extends helpers for common patterns: string/int/bool validation with configurable expected status, numeric path IDs, required query parameters, and extra-field (parameter injection) checks.
- Keeps test behavior consistent while reducing duplication and making it easier to add new tests.

Details:
- Updates `validationHelpers.js`:
   - Extends validateString, validateInt, and validateBool to accept an expectedStatus parameter.
   - Adds validateNumericIdInPath, expectValidNumericIdInPath, and validateRequiredQueryParams helpers for common ID and query validation patterns.
- Refactors the following tests to use the helpers:
`compareReadingsParamsTest.js`
`conversionsParamsTest.js`
`groupsParamsTest.js`
`mapsParamsTest.js`
`readingsParamsTest.js`
`unitAddParamsTest.js`

Fixes #1572 

Author: Kaulan Serzhanuly

(In general, OED likes to have at least one issue associated with each pull request. Replace [issue] with the OED GitHub issue number. In the preview you will see an issue description if you hover over that number. You can create one yourself before doing this pull request. This is where details are normally given on what is being addressed. Note you should not use the word "Fixes" if it does not completely address the issue since the issue would automatically be closed on merging the pull request. In that case use "Partly Addresses #[issue].)

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [ ] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

(Describe any issues that remain or work that should still be done.)

### CODE REVIEW

**Comment by huss** [comment]
@KaulanSerzhanuly I see you checked the box for having completed the CLA. However, I did not find it in the OED records. If you have under a different GitHub ID or think the OED records are off then let me know. Otherwise, could you please fill it out.

**Comment by KaulanSerzhanuly** [comment]
Hello I did sign everything tho, it was under this accounts ID. This accounts email is kaulan.2004@gmail.com and name to this account is KaulanSerzhanuly. Just in case i have resubmited it

**Comment by KaulanSerzhanuly** [comment]
I forgot to tag you sir @huss

**Comment by KaulanSerzhanuly** [comment]
@huss Hello again i resubmited OED record. Could you please check it again?

File: src/server/test/util/validationHelpers.js [hunk 1]
```diff
@@ -64,26 +64,36 @@ async function validateMinMaxRelation({ endpoint, basePayload }) {
  * @param minLength the minimum length allowed for the string (default: 1)
  * @param maxLength the maximum length allowed for the string (default: 255)
  * @param enumValues optional array of valid enum values to test against
+ * @param expectedStatus expected HTTP status code(s) (default: 400)
```
Comment (huss) [inline src/server/test/util/validationHelpers.js]: It might be nice to add to the 400 that it is HTTP_CODE.BAD_REQUEST for developers since that is actually used. Applies to other, similar functions. I also see that there were ones you did not touch with the same consideration. You are welcome to update those.

File: src/server/test/util/validationHelpers.js [hunk 2]
```diff
@@ -224,6 +236,93 @@ async function expectValidCommaSeparatedIds({
 	}
 }
 
+/**
+ * Validates that endpoints reject invalid single numeric IDs in path parameters.
+ * Convenience wrapper around validateCommaSeparatedIdPatterns for routes like
```
Comment (huss) [inline src/server/test/util/validationHelpers.js]: This says it is a convenience wrapper around the other function but it does not call it. There is a high overlap of the code in the two functions. I'm wondering if it is possible to combine them into one with appropriate defaults or get one to call the other. Do you think this could be done to allow more code reuse rather than duplication? The same idea may apply to the next function too.

File: src/server/test/util/validationHelpers.js [hunk 3]
```diff
@@ -224,6 +236,93 @@ async function expectValidCommaSeparatedIds({
...
+
+		if (method.toLowerCase() === 'get') {
+			request = request.query(query);
+		} else {
+			request = request.send(query);
+		}
+
+		const res = await request;
+		const expectedStatuses = Array.isArray(expectedStatus) ? expectedStatus : [expectedStatus];
+		expect(expectedStatuses).to.include(res.status);
```
Comment (huss) [inline src/server/test/util/validationHelpers.js:300]: If I understand this code correctly, it allows for single or array values to be passed. If this is correct then I wondering about the parameter name/description.

File: src/server/test/util/validationHelpers.js [hunk 4]
```diff
@@ -224,6 +236,93 @@ async function expectValidCommaSeparatedIds({
...
+
+		let request = chai.request(app)[method](endpoint);
+		if (method.toLowerCase() === 'get') {
+			request = request.query(incompleteQuery);
+		} else {
+			request = request.send(incompleteQuery);
+		}
+
+		const res = await request;
+		const expectedStatuses = Array.isArray(expectedStatus) ? expectedStatus : [expectedStatus];
```
Comment (huss) [inline src/server/test/util/validationHelpers.js:352]: See comment in validateNumericIdInPath.

File: src/server/test/routes/compareReadingsParamsTest.js [hunk 5]
```diff
@@ -75,18 +76,12 @@ mocha.describe('Compare Readings Parameter Validation', () => {
...
-						.get(`${BASE_METER_ENDPOINT}/1`)
-						.query(incompleteQuery);
-
-					expect(res.status).to.equal(HTTP_CODE.BAD_REQUEST);
-				}
+				await validateRequiredQueryParams({
+					endpoint: `${BASE_METER_ENDPOINT}/1`,
+					baseQuery: validQuery,
+					requiredParams: ['curr_start', 'curr_end', 'shift', 'graphicUnitId'],
+					expectedStatus: HTTP_CODE.BAD_REQUEST
```
Comment (huss) [inline src/server/test/routes/compareReadingsParamsTest.js]: There is nothing wrong with this but here you give a value that is the same as the default but for another value you do not. Is there a reason?

File: src/server/test/routes/conversionsParamsTest.js [hunk 6]
```diff
@@ -60,33 +60,23 @@ mocha.describe('Conversions Parameter Validation', () => {
...
-				field: 'sourceId',
-				invalidValue: -1,
-				endpoint: EDIT_ENDPOINT,
-				basePayload: baseConversionData,
-				expectedStatus: HTTP_CODE.FORBIDDEN
-			});
-
-			// Test negative destinationId (minimum: 1)
-			await testInvalidField({
+			await validateInt({
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js:71]: I like the centralization of testing of the int values. One small change that may not make any practical difference is it used to test 0 and -1. I think the idea was that one is negative and zero is special so it might slip through. What would you think of still testing min of -1 but add that if min > 1 then test 0 and if min > 0 then test -1 (in new function for this)?

File: src/server/test/routes/conversionsParamsTest.js [hunk 7]
```diff
@@ -60,33 +60,23 @@ mocha.describe('Conversions Parameter Validation', () => {
...
 				field: 'destinationId',
-				invalidValue: 0,
 				endpoint: EDIT_ENDPOINT,
 				basePayload: baseConversionData,
+				min: 1,
 				expectedStatus: HTTP_CODE.FORBIDDEN
 			});
 
-			// Test oversized integers (exceeds MAX_SAFE_INTEGER)
+			// Test oversized integers (exceeds MAX_SAFE_INTEGER) - not covered by validateInt
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js]: If this test is a good idea then could it be rolled into the standard int test? I cannot think of any case where this would be allowed. Now, if there is a max specified then it probably is not needed but will not hurt in that case.

File: src/server/test/routes/conversionsParamsTest.js [hunk 8]
```diff
@@ -95,18 +85,9 @@ mocha.describe('Conversions Parameter Validation', () => {
...
-			// Test non-integer values
-			await testInvalidField({
-				field: 'sourceId',
-				invalidValue: 'not_a_number',
-				endpoint: EDIT_ENDPOINT,
-				basePayload: baseConversionData,
-				expectedStatus: HTTP_CODE.FORBIDDEN
-			});
-
+			// Test float instead of integer
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js]: I did not search all the tests but could this be added to standard int tests?

File: src/server/test/routes/conversionsParamsTest.js [hunk 9]
```diff
@@ -146,26 +127,20 @@ mocha.describe('Conversions Parameter Validation', () => {
...
-			for (const invalidValue of invalidBooleanValues) {
-				await testInvalidField({
-					field: 'bidirectional',
-					invalidValue: invalidValue,
-					endpoint: EDIT_ENDPOINT,
-					basePayload: baseConversionData,
-					expectedStatus: HTTP_CODE.FORBIDDEN
-				});
-			}
+			await validateBool({
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js:112]: I like the centralization. It seems the new function does not test as many possible invalid values as this used to do. I'm inclined to do all the old cases since they might happen in code.

File: src/server/test/routes/conversionsParamsTest.js [hunk 10]
```diff
@@ -146,26 +127,20 @@ mocha.describe('Conversions Parameter Validation', () => {
...
+				field: 'bidirectional',
+				endpoint: EDIT_ENDPOINT,
+				basePayload: baseConversionData,
+				expectedStatus: HTTP_CODE.FORBIDDEN
+			});
 
-			// Test valid boolean values
+			// Test valid boolean values pass validation (fail auth)
 			const validBooleanValues = [true, false];
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js]: I'm wondering if the new boolean function should test this to centralize. I see other files did not include this check but I'm unclear on when it would be an issue. I think for admin ones this is fine and for other routes one might be able to check if the return code isn't for bad params but that would need to be verified.

File: src/server/test/routes/conversionsParamsTest.js [hunk 11]
```diff
@@ -387,39 +358,19 @@ mocha.describe('Conversions Parameter Validation', () => {
...
-				field: 'destinationId',
-				invalidValue: 0,
-				endpoint: DELETE_ENDPOINT,
-				basePayload: baseDeleteData,
-				expectedStatus: HTTP_CODE.FORBIDDEN
-			});
-
-			// Test non-integer ID
-			await testInvalidField({
+			await validateInt({
```
Comment (huss) [inline src/server/test/routes/conversionsParamsTest.js:334]: See above about centralizing the int test. I think this might cover this case too.

File: src/server/test/routes/groupsParamsTest.js [hunk 12]
```diff
@@ -19,97 +21,79 @@ const {
...
-			for (const invalidId of invalidIds) {
-				const res = await chai.request(app)
-					.get(`/api/groups/deep/groups/${invalidId}`);
-
-				// TODO: Some invalid IDs like 'abc' return 200 instead of expected 400
-				// This suggests path parameter validation may not be working as intended
-				// or Express is interpreting these as valid somehow. Needs investigation.
-				expect([HTTP_CODE.OK, HTTP_CODE.BAD_REQUEST]).to.include(res.status);
-			}
+			// TODO: Some invalid IDs may return 200 - path param validation may need investigation
```
Comment (huss) [inline src/server/test/routes/groupsParamsTest.js]: Could you help me understand the status of this TODO?

File: src/server/test/routes/groupsParamsTest.js [hunk 13]
```diff
@@ -19,97 +21,79 @@ const {
...
-			for (const invalidId of invalidIds) {
-				const res = await chai.request(app)
-					.get(`/api/groups/parents/${invalidId}`);
+		const BASE_ENDPOINT = '/api/groups/parents';
 
-				// Should return 400 for validation error or 200 if somehow valid
-				expect([HTTP_CODE.OK, HTTP_CODE.BAD_REQUEST]).to.include(res.status);
-			}
+		mocha.it('should validate group_id parameter', async () => {
+			await validateNumericIdInPath({
```
Comment (huss) [inline src/server/test/routes/groupsParamsTest.js]: It seems that this test and the next one are repeated across paths. Could they be centralized?

File: src/server/test/routes/mapsParamsTest.js [hunk 14]
```diff
@@ -30,38 +35,27 @@ mocha.describe('Maps Parameter Validation', () => {
 		const BASE_ENDPOINT = '/api/maps';
 
 		mocha.it('should accept valid map ID', async () => {
-			const res = await chai.request(app)
-				.get(`${BASE_ENDPOINT}/1`);
-
-			// Should return 200 or 500 (DB error) - no auth required for reading
-			expect([HTTP_CODE.OK, HTTP_CODE.INTERNAL_SERVER_ERROR]).to.include(res.status);
+			await expectValidNumericIdInPath({
```
Comment (huss) [inline src/server/test/routes/mapsParamsTest.js:38]: I'm noting that this is an example of similar tests from other files that I'm thinking about centralization.

File: src/server/test/routes/mapsParamsTest.js [hunk 15]
```diff
@@ -329,16 +314,12 @@ mocha.describe('Maps Parameter Validation', () => {
...
-				...baseDeleteData,
-				maliciousField: 'injection_attempt'
-			};
-
-			const res = await chai.request(app)
-				.post(DELETE_ENDPOINT)
-				.send(payloadWithExtra);
-
-			expect(res.status).to.equal(HTTP_CODE.FORBIDDEN);
+			await validateNoExtraFields({
```
Comment (huss) [inline src/server/test/routes/mapsParamsTest.js:317]: I'm not going to mark every one but here is one where I wonder about a wrapper function to set these values and then call the validate function as one or more params are the same across tests.

File: src/server/test/routes/readingsParamsTest.js [hunk 16]
```diff
@@ -7,49 +7,51 @@
...
+	validateRequiredQueryParams,
+	validateNumericIdInPath
+} = require('../util/validationHelpers');
 
 mocha.describe('Readings Route Parameter Validation', () => {
 
 	const LINE_COUNT_BASE_ENDPOINT = '/api/readings/line/count/meters';
 	const RAW_READINGS_BASE_ENDPOINT = '/api/readings/line/raw/meter';
 
+	const lineCountValidQuery = { timeInterval: '2020-01-01T00:00:00.000Z_2020-01-02T00:00:00.000Z' };
```
Comment (huss) [inline src/server/test/routes/readingsParamsTest.js]: This same param shows up 7 other times in the file. Could one central value to used?

File: src/server/test/routes/readingsParamsTest.js [hunk 17]
```diff
@@ -105,65 +107,38 @@ mocha.describe('Readings Route Parameter Validation', () => {
...
+					validValues: ['1'],
+					query: rawReadingsValidQuery,
+					expectedStatuses: [HTTP_CODE.OK, HTTP_CODE.BAD_REQUEST, HTTP_CODE.NOT_FOUND, HTTP_CODE.INTERNAL_SERVER_ERROR]
+				});
+			});
+
+			mocha.it('should reject invalid meter_id patterns', async () => {
+				await validateNumericIdInPath({
+					baseEndpoint: RAW_READINGS_BASE_ENDPOINT,
+					invalidValues: ['not_a_number', '0', '-1', '1.5', '2147483648', encodeURIComponent("1'; DROP TABLE readings; --"), '99999999999999999999999999999999'],
```
Comment (huss) [inline src/server/test/routes/readingsParamsTest.js]: Small point but maybe 2147483648 should relate to the JS constant for max value and 99999999999999999999999999999999 could be done via repeat as elsewhere.

File: src/server/test/routes/unitAddParamsTest.js [hunk 18]
```diff
@@ -9,7 +9,9 @@ const { chai, mocha, app } = require('../common');
 const { HTTP_CODE } = require('../../util/readingsUtils');
```
Comment (huss) [inline src/server/test/routes/unitAddParamsTest.js:9]: It seems some of the tests here might relate to general tests from other files such as int and float. Some seem unique to here but may apply more generally. I know this comment is vague so let me know if you want to discuss more.

File: src/server/test/routes/unitAddParamsTest.js [hunk 19]
```diff
@@ -62,59 +64,26 @@ mocha.describe('Units Add Parameter Validation', () => {
...
-				field: 'displayable',
-				invalidValue: '',
-				endpoint: ADD_ENDPOINT,
-				basePayload: baseUnitData,
-				expectedStatus: HTTP_CODE.FORBIDDEN
-			});
+			const stringFields = [
+				{ field: 'name', maxLength: STRING_SHORT_MAX_LENGTH },
+				{ field: 'identifier', maxLength: STRING_GENERAL_MAX_LENGTH },
+				{ field: 'unitRepresent', maxLength: STRING_GENERAL_MAX_LENGTH },
```
Comment (huss) [inline src/server/test/routes/unitAddParamsTest.js]: I think starting with this string, these are actually enums (or frozen objects in JS) that specify the valid values. I wonder if they should be tested as such.

**Review by huss — CHANGES_REQUESTED** [review]
@KaulanSerzhanuly First, I'm sorry this took a few days while my efforts needed to be elsewhere. Second, thank you for this work. Overall, I like these changes. It caused me to think in a deeper way about possible centralization and I've made a number of comments about this. I want to note that I did not carefully think them through so it may be the case that it does not apply across all routes. It is also possible my modest comments are not clear enough. Please feel free to put in your thoughts and ask me about anything. I'm always here to help.

**Comment by KaulanSerzhanuly** [comment]
Thanks for the thorough review and for flagging that some ideas might not fit every route—I’ll call those out where I left them for later.
I pushed updates that: document defaults as `HTTP_CODE.BAD_REQUEST`, merge the path-ID helpers via a shared implementation + clarify `expectedStatus` as number | number[], extend `validateInt` / `validateBool` (including the edge cases you mentioned), centralize the duplicated group_id tests and the readings `timeInterval`, treat unit enum-like fields with enumValues in `validateString`, and drop redundant expectedStatus where it matched the default. Happy to discuss anything you’d still like generalized.

**Comment by KaulanSerzhanuly** [comment]
@huss Please check out the lastest changes I pushed

File: src/server/test/util/validationHelpers.js (hunk 2 above)
Comment (KaulanSerzhanuly) [reply src/server/test/util/validationHelpers.js]: Done - both call shared validatePathIdSegmentPatterns; comments updated so it’s not “wrapper” without delegation.

File: src/server/test/routes/compareReadingsParamsTest.js (hunk 5 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/compareReadingsParamsTest.js]: No strong reason - it was inconsistent. I removed the redundant expectedStatus where it matched the helper default so it’s the same style everywhere.

File: src/server/test/routes/conversionsParamsTest.js (hunk 6 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/conversionsParamsTest.js:71]: Good idea. validateInt now also sends -1 when min > 0 and 0 when min > 1 (in addition to min - 1), so those edge cases stay covered without duplicating them per route.

File: src/server/test/routes/conversionsParamsTest.js (hunk 7 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/conversionsParamsTest.js]: Done. validateInt now includes Number.MAX_SAFE_INTEGER + 1, and the conversions tests drop the extra one-off assert for that.

File: src/server/test/routes/groupsParamsTest.js (hunk 12 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/groupsParamsTest.js]: Those routes don’t consistently reject malformed group_id in the path today - some cases still return 200 until path validation is tightened. The test now documents that with a short comment, and still allows [OK, BAD_REQUEST] so we’re not asserting behavior the server doesn’t guarantee yet.

File: src/server/test/routes/mapsParamsTest.js (hunk 15 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/mapsParamsTest.js:317]: Reasonable. I didn’t add that wrapper here to keep the diff focused; happy to add a local helper (or a tiny test util) in a follow-up if we standardize on the same forbidden + base payload shape in several tests.

File: src/server/test/routes/readingsParamsTest.js (hunk 16 above)
Comment (KaulanSerzhanuly) [reply src/server/test/routes/readingsParamsTest.js]: Done. There’s a single module-level READINGS_LINE_TIME_INTERVAL (and queries derive from it) so the literal isn’t copy-pasted.

**Comment by huss** [comment]
> @huss Please check out the lastest changes I pushed

Oops! I lost track of this. I plan to look at it in the near future.

File: src/server/test/routes/groupsParamsTest.js (hunk 12 above)
Comment (huss) [reply src/server/test/routes/groupsParamsTest.js]: This has been a long-standing idea in my mind and this pushed it over the top. I have created issue #1592 to fix this validation. It is linked to this PR to show your comment.

Comment (huss) [reply src/server/test/routes/groupsParamsTest.js]: This is resolved but leaving open for others to see.

**Review by huss — APPROVED** [review]
@KaulanSerzhanuly Thank you for carefully and thoroughly addressing my previous comments. There are a couple unresolved but I view them as future work. I don't want perfection to be the impediment to progress. You fixed all the important cases (and then more) and this addresses creating general test code for routes. If you want to do anything else on this or something else then you are welcome to contact me. Congratulations on your first accepted pull request to OED.
